pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,452 @@
+type CGroupDiskStaller struct {
+	GenericFailure
+}
+
+func MakeCgroupDiskStaller(clusterName string, l *logger.Logger, secure bool) (FailureMode, error) {
+	c, err := roachprod.GetClusterFromCache(l, clusterName, install.SecureOption(secure))
+	if err != nil {
+		return nil, err
+	}
+	genericFailure := GenericFailure{c: c, runTitle: CgroupsDiskStallName}
+	return &CGroupDiskStaller{GenericFailure: genericFailure}, nil
+}
+
+const CgroupsDiskStallName = "cgroup-disk-stall"
+
+func registerCgroupDiskStall(r *FailureRegistry) {
+	r.add(CgroupsDiskStallName, DiskStallArgs{}, MakeCgroupDiskStaller)
+}
+
+func (s *CGroupDiskStaller) Description() string {
+	return CgroupsDiskStallName
+}
+
+type DiskStallArgs struct {
+	StallLogs   bool
+	StallReads  bool
+	StallWrites bool
+	// If true, allow the failure mode to restart nodes as needed. E.g. dmsetup requires
+	// the cockroach process to not be running to properly setup. If RestartNodes is true,
+	// then the failure mode will restart the cluster for the user.
+	RestartNodes bool
+	// Throughput is the bytes per second to throttle reads/writes to. If unset, will
+	// stall reads/write completely. Supported only for cgroup disk staller, dmsetup
+	// only supports fully stalling reads/writes.
+	Throughput int
+	Nodes      install.Nodes
+}
+
+func (s *CGroupDiskStaller) Setup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	diskStallArgs := args.(DiskStallArgs)
+
+	if diskStallArgs.StallLogs {
+		if err := s.Run(ctx, l, s.c.Nodes, "mkdir -p {store-dir}/logs"); err != nil {
+			return err
+		}
+		if err := s.Run(ctx, l, s.c.Nodes, "rm -f logs && ln -s {store-dir}/logs logs || true"); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (s *CGroupDiskStaller) Cleanup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	stallType := []bandwidthType{readBandwidth, writeBandwidth}
+	nodes := args.(DiskStallArgs).Nodes
+
+	// Setting cgroup limits is idempotent so attempt to unlimit reads/writes in case
+	// something went wrong in Restore.
+	err := s.setThroughput(ctx, l, stallType, throughput{limited: false}, nodes, cockroachIOController)
+	if err != nil {
+		l.PrintfCtx(ctx, "error unstalling the disk; stumbling on: %v", err)
+	}
+	return nil
+}
+
+func getStallType(diskStallArgs DiskStallArgs) ([]bandwidthType, error) {
+	var stallType []bandwidthType
+	if diskStallArgs.StallWrites {
+		stallType = []bandwidthType{writeBandwidth}
+	}
+	if diskStallArgs.StallReads {
+		stallType = append(stallType, readBandwidth)
+	}
+	if len(stallType) == 0 {
+		return nil, errors.New("at least one of reads or writes must be stalled")
+	}
+	return stallType, nil
+}
+
+func (s *CGroupDiskStaller) Inject(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	diskStallArgs := args.(DiskStallArgs)
+	stallType, err := getStallType(diskStallArgs)
+	if err != nil {
+		return err
+	}
+
+	// N.B. Although the cgroupsv2 documentation states that "limits are in the range [0, max]",
```

Comment:
Mystery solved after spending way too long reading reading the source code for the wrong distribution of Linux 🤣 Well, kinda 😅 . I assume 0 isn't allowed to avoid division by 0 when it's calculating the allowed bytes. No clue why 1 isn't allowed.